Snowflake.Data/Core/Converter/JsonToStructuredTypeConverter.cs

```diff
@@ -0,0 +1,423 @@
+
+                return ConvertToUnstructuredType(fieldMetadata, valueType, fieldValue);
+            }
+            catch (Exception e)
+            {
+                if (e is SnowflakeDbException)
+                    throw;
+                throw new SnowflakeDbException(SFError.STRUCTURED_TYPE_READ_ERROR, $"when reading '{fieldMetadata.name}' property", e.Message);
+            }
+        }
+
+        private static Type GetNestedType(Type type)
+        {
+            if (type.IsArray)
+            {
+                return type.GetElementType();
+            }
+            if (IsListType(type) || IsIListType(type))
+            {
+                return type.GenericTypeArguments[0];
+            }
+            throw new Exception("neither array nor list");
+        }
+
+        private static Type[] GetMapKeyValueTypes(Type type)
+        {
+            var genericParamWithTwoArguments = type.IsGenericType && type.GenericTypeArguments.Length == 2;
+            if (!genericParamWithTwoArguments)
+            {
+                throw new Exception("Could not get key and value types");
+            }
+            return type.GenericTypeArguments;
+        }
+
+        private static Type MakeArrayType(Type type, Type elementType)
+        {
+            if (type.IsArray)
+            {
+                return type;
+            }
+            if (IsListType(type) || IsIListType(type))
+            {
+                return elementType.MakeArrayType();
+            }
+            throw new Exception("Neither array nor list");
+        }
+
+        private static bool IsListType(Type type) => type.IsGenericType && type.GetGenericTypeDefinition() == typeof(List<>);
+
+        private static bool IsIListType(Type type) => type.IsGenericType && type.GetGenericTypeDefinition() == typeof(IList<>);
+
+        private static bool IsStructured(FieldMetadata fieldMetadata) =>
+            fieldMetadata.fields != null && fieldMetadata.fields.Count > 0;
+
+        private static bool IsObjectMetadata(FieldMetadata fieldMetadata) =>
+            SFDataType.OBJECT.ToString().Equals(fieldMetadata.type, StringComparison.OrdinalIgnoreCase);
+
+        private static bool IsTextMetadata(FieldMetadata fieldMetadata) =>
+            SFDataType.TEXT.ToString().Equals(fieldMetadata.type, StringComparison.OrdinalIgnoreCase);
+
+        private static bool IsFixedMetadata(FieldMetadata fieldMetadata) =>
+            SFDataType.FIXED.ToString().Equals(fieldMetadata.type, StringComparison.OrdinalIgnoreCase);
+
+        private static bool IsTimestampNtz(FieldMetadata fieldMetadata) =>
+            SFDataType.TIMESTAMP_NTZ.ToString().Equals(fieldMetadata.type, StringComparison.OrdinalIgnoreCase);
+
+        private static bool IsTimestampLtz(FieldMetadata fieldMetadata) =>
+            SFDataType.TIMESTAMP_LTZ.ToString().Equals(fieldMetadata.type, StringComparison.OrdinalIgnoreCase);
+
+        private static bool IsTimestampTz(FieldMetadata fieldMetadata) =>
```

Comment:
Why there sometimes is `Metadata` suffix and sometimes not?

Reply:
fixed